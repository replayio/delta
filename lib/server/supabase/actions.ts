import {
  PostgrestResponse,
  PostgrestSingleResponse,
} from "@supabase/supabase-js";
import { Action, ActionId, JobId } from "../../types";
import { retryOnError, supabase } from "./supabase";

export async function getAction(
  actionId: ActionId
): Promise<PostgrestSingleResponse<Action>> {
  return retryOnError(() =>
    supabase.from("Actions").select("*").eq("id", actionId).single()
  );
}

export async function getActionsForJob(
  jobId: JobId
): Promise<PostgrestResponse<Action>> {
  return retryOnError(() =>
    supabase
      .from("Actions")
      .select("*")
      .eq("job_id", jobId)
      .order("created_at", { ascending: false })
  );
}

export async function updateAction(
  actionId: ActionId,
  action: Partial<Action>
): Promise<PostgrestSingleResponse<Action>> {
  return retryOnError(() =>
    supabase.from("Actions").update(action).eq("id", actionId).single()
  );
}
