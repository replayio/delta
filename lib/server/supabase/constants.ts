// https://postgrest.org/en/latest/references/errors.html
export const POSTGRESQL_ERROR_CODES = {
  UNIQUENESS_VIOLATION: 23505,
};
