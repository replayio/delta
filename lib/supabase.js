import createClient from "./initServerSupabase";
const { createHash } = require("crypto");

const supabase = createClient();

export async function getProjectIdFromActionId(actionId) {
  const actionRecord = await supabase
    .from("Actions")
    .select("*")
    .eq("id", actionId);
  console.log(actionRecord);
  const projectId = actionRecord.data[0].project_id;
  return projectId;
}

export async function getSnapshotFromBranch(image, projectId, branchName) {
  const { branch } = await getBranchFromProject(projectId, branchName);

  if (!branch) {
    return { error: "Branch not found" };
  }

  const response = await supabase
    .from("Snapshots")
    .select("*, Actions(branch_id)")
    .eq("file", image.file)
    .eq("Actions.branch_id", branch.id)
    .limit(1);

  if (response.error) {
    console.log("getSnapshotFromBranch error", response.error);
    return { error: response.error };
  }

  return { snapshot: response.data?.[0] };
}

export async function getSnapshotsFromBranch(projectId, branchName) {
  const { branch } = await getBranchFromProject(projectId, branchName);

  if (!branch) {
    return { error: "Branch not found" };
  }

  const response = await supabase
    .from("Snapshots")
    .select("*, Actions(branch_id)")
    .eq("Actions.branch_id", branch.id);

  if (response.error) {
    console.log("getSnapshotFromBranch error", response.error);
    return { error: response.error };
  }

  return { snapshots: response.data };
}

export async function insertSnapshot(
  branchName,
  projectId,
  image,
  status,
  primary_changed,
  action_changed
) {
  const { branch } = await getBranchFromProject(projectId, branchName);
  const { action } = await getAction(branch.id);
  const sha = createHash("sha256").update(image.content).digest("hex");

  return supabase.from("Snapshots").insert({
    sha,
    action_id: action.id,
    path: `${projectId}/${sha}.png`,
    file: image.file,
    status,
    action_changed,
    primary_changed,
  });
}

export async function getProject(projectId) {
  const projectResponse = await supabase
    .from("Projects")
    .select("*")
    .eq("id", projectId)
    .order("created_at", { ascending: false })
    .limit(1);

  if (projectResponse.error) {
    console.log("getProject error", projectResponse.error);
    return { error: projectResponse.error };
  }

  return { project: projectResponse.data[0] };
}

export async function getProjectFromRepo(repository, organization) {
  const projectResponse = await supabase
    .from("Projects")
    .select("*")
    .eq("organization", organization)
    .eq("repository", repository)
    .limit(1);

  if (projectResponse.error) {
    console.log("error", projectResponse.error);
    return { error: projectResponse.error };
  }

  return projectResponse.data[0];
}

export async function getBranchFromProject(projectId, branch) {
  const res = await supabase
    .from("Branches")
    .select("*")
    .eq("project_id", projectId)
    .eq("name", branch)
    .limit(1);

  if (res.error) {
    return { error: res.error };
  }

  return { branch: res.data[0] };
}

export async function getAction(branch_id) {
  const resp = await supabase
    .from("Actions")
    .select("*")
    .eq("branch_id", branch_id)
    .order("created_at", { ascending: false })
    .limit(1);

  if (resp.error) {
    return { error: resp.error };
  }

  return { action: resp.data[0] };
}
