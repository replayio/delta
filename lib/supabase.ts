import {
  PostgrestResponse,
  PostgrestSingleResponse,
} from "@supabase/supabase-js";
import createClient from "./initServerSupabase";
const { createHash } = require("crypto");

const supabase = createClient();

export type Project = {
  id: string;
  repository: string;
  organization: string;
  created_at: string;
  primary_branch: string;
};

export type Branch = {
  id: string;
  name: string;
  project_id: string;
  pr_url?: string;
  pr_title?: string;
  pr_number?: string;
  created_at: string;
  check_id: string;
  status: string;
};

export type Action = {
  id: string;
  branch_id: string;
  created_at: string;
  run_id: string;
  head_sha: string;
  actor: string;
};

export type Snapshot = {
  id: string;
  sha: string;
  action_id: string;
  path: string;
  file: string;
  status: string;
  action_changed: boolean;
  primary_changed: boolean;
  created_at: string;
};

type ResponseError = {
  error: string;
  data: null;
};

const createError = (error: string): ResponseError => ({ error, data: null });

export async function getSnapshotFromBranch(
  image: { file: string; content: string },
  projectId: string,
  branchName: string
): Promise<ResponseError | PostgrestSingleResponse<Snapshot>> {
  const branch = await getBranchFromProject(projectId, branchName);

  if (branch.error) {
    return createError("Branch not found");
  }

  return supabase
    .from("Snapshots")
    .select("*, Actions(branch_id)")
    .eq("file", image.file)
    .eq("Actions.branch_id", branch.data.id)
    .single();
}

export async function getSnapshotsFromBranch(
  projectId: string,
  branchName: string
): Promise<ResponseError | PostgrestResponse<Snapshot>> {
  const branch = await getBranchFromProject(projectId, branchName);

  if (branch.error) {
    return createError("Branch not found");
  }

  const actions = await getActionsFromBranch(branch.data.id);

  if (actions.error) {
    return createError("Action not found");
  }

  for (const action of actions.data) {
    console.log(
      `getSnapshotsFromBranch(${projectId}, ${branchName})`,
      action.id
    );

    const snapshots = await supabase
      .from("Snapshots")
      .select("*")
      .eq("action_id", action.id);

    if (snapshots.error) {
      continue;
    }
    if (snapshots.data.length > 0) {
      return snapshots;
    }
  }
}

export async function insertSnapshot(
  branchName,
  projectId,
  image,
  status,
  primary_changed,
  action_changed = false
): Promise<ResponseError | PostgrestSingleResponse<Snapshot>> {
  const branch = await getBranchFromProject(projectId, branchName);
  if (branch.error) {
    return createError(`Branch not found for ${branchName}`);
  }
  const action = await getActionFromBranch(branch.data.id);
  if (action.error) {
    return createError(
      `Action not found for ${branchName} and ${
        branch.data.id
      }: ${JSON.stringify(action.error)}`
    );
  }

  const sha = createHash("sha256").update(image.content).digest("hex");

  return supabase
    .from("Snapshots")
    .insert({
      sha,
      action_id: action.data.id,
      path: `${projectId}/${sha}.png`,
      file: image.file,
      status,
      action_changed,
      primary_changed,
    })
    .single();
}

export async function getProject(
  projectId
): Promise<PostgrestSingleResponse<Project>> {
  return supabase
    .from("Projects")
    .select("*")
    .eq("id", projectId)
    .order("created_at", { ascending: false })
    .single();
}

export async function getProjectByShort(
  projectShort: string
): Promise<PostgrestSingleResponse<Project>> {
  return supabase
    .from("Projects")
    .select("*")
    .eq("short", projectShort)
    .single();
}

export async function getPublicProjects(): Promise<
  PostgrestResponse<Project[]>
> {
  return supabase
    .from("Projects")
    .select("*")
    .eq("public", true)
    .order("created_at", { ascending: false });
}

export async function getProjectFromRepo(
  repository: string,
  organization: string
): Promise<PostgrestSingleResponse<Project>> {
  return supabase
    .from("Projects")
    .select("*")
    .eq("organization", organization)
    .eq("repository", repository)
    .single();
}

export async function getBranchFromProject(
  projectId: string,
  branch: string
): Promise<PostgrestSingleResponse<Branch>> {
  return supabase
    .from("Branches")
    .select("*")
    .eq("project_id", projectId)
    .eq("name", branch)
    .single();
}

export async function updateBranch(branch: Branch, data: Partial<Branch>) {
  return supabase.from("Branches").update(data).eq("id", branch.id).single();
}

export async function getActionFromBranch(
  branch_id: string
): Promise<PostgrestSingleResponse<Action>> {
  return supabase
    .from("Actions")
    .select("*")
    .eq("branch_id", branch_id)
    .order("created_at", { ascending: false })
    .limit(1)
    .single();
}

export async function getActionsFromBranch(
  branch_id: string
): Promise<PostgrestResponse<Action>> {
  return supabase
    .from("Actions")
    .select("*")
    .eq("branch_id", branch_id)
    .order("created_at", { ascending: false });
}

export async function getAction(
  actionId: string
): Promise<PostgrestSingleResponse<Action>> {
  return supabase.from("Actions").select("*").eq("id", actionId).single();
}
