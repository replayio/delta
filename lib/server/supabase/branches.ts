import {
  PostgrestResponse,
  PostgrestSingleResponse,
} from "@supabase/supabase-js";
import { Branch, BranchId, ProjectId } from "../../types";
import { retryOnError, supabase } from "./supabase";

export async function getBranchForProject(
  projectId: ProjectId,
  branchName: string
): Promise<PostgrestSingleResponse<Branch>> {
  return retryOnError(() =>
    supabase
      .from("Branches")
      .select("*")
      .eq("project_id", projectId)
      .eq("name", `${branchName}`)
      .order("created_at", { ascending: false })
      .limit(1)
      .single()
  );
}

export async function getBranchForPullRequest(
  projectId: ProjectId,
  prNumber: number
): Promise<PostgrestSingleResponse<Branch>> {
  return retryOnError(() =>
    supabase
      .from("Branches")
      .select("*")
      .eq("project_id", projectId)
      .eq("pr_number", prNumber)
      .single()
  );
}

export async function getBranch(
  branchId: BranchId
): Promise<PostgrestSingleResponse<Branch>> {
  return retryOnError(() =>
    supabase.from("Branches").select("*").eq("id", branchId).single()
  );
}

export async function getBranchByName(
  name: string,
  status: string = "open"
): Promise<PostgrestSingleResponse<Branch>> {
  return retryOnError(() =>
    supabase
      .from("Branches")
      .select("*")
      .eq("name", name)
      .eq("status", status)
      .order("created_at", { ascending: false })
      .limit(1)
      .single()
  );
}

export async function getBranchesForProject(
  projectId: ProjectId
): Promise<PostgrestResponse<Branch>> {
  return retryOnError(() =>
    supabase.from("Branches").select("*").eq("project_id", projectId)
  );
}

export async function updateBranch(branchId: BranchId, data: Partial<Branch>) {
  return retryOnError(() =>
    supabase.from("Branches").update(data).eq("id", branchId).single()
  );
}

export async function insertBranch(
  newBranch: Partial<Branch>
): Promise<PostgrestSingleResponse<Branch>> {
  return retryOnError(() =>
    supabase.from("Branches").insert(newBranch).single()
  );
}
