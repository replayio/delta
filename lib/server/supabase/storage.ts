import createClient from "../../initServerSupabase";
import { createHash } from "crypto";

import type { FileObject } from "@supabase/storage-js";
import { maybeRetry, retryOnError } from "./supabase";

const supabase = createClient();
const Buffer = require("buffer").Buffer;

export type StoredSnapshot = {
  path: string;
  sha: string;
};

type UploadResult = {
  data: StoredSnapshot | null;
  error: string | null;
};

export async function uploadSnapshot(
  content: string | Buffer,
  projectId: string
): Promise<UploadResult> {
  const sha = createHash("sha256").update(content).digest("hex");
  const path = `${projectId}/${sha}.png`;

  if (typeof content === "string") {
    content = Buffer.from(content, "base64");
  }

  const res = await maybeRetry(
    () =>
      supabase.storage.from("snapshots").upload(path, content, {
        contentType: "image/png",
      }),
    (result) =>
      !!result.error && result.error.message !== "The resource already exists"
  );

  if (res.error) {
    return { error: (res.error as any).error, data: { sha, path } };
  }

  return { data: { path, sha }, error: null };
}

export async function downloadSnapshot(
  path: string
): Promise<{ error: Error | null; data: string | null }> {
  const { data, error } = await retryOnError(() =>
    supabase.storage.from("snapshots").download(path)
  );

  if (data == null || error) {
    return { error, data: null };
  }

  const fileBuffer = Buffer.from(await data.arrayBuffer());

  const image = fileBuffer.toString("base64");
  return { data: image, error: null };
}

export async function listSnapshots(
  projectId: string
): Promise<{ data: FileObject[] | null; error: string | null }> {
  const { data, error } = await retryOnError(() =>
    supabase.storage.from("snapshots").list(projectId, {
      limit: 100,
    })
  );

  if (error) {
    return { error: error as any, data: null };
  }

  return { data, error: null };
}

export async function listCorruptedSnapshots(
  projectId: string
): Promise<any[]> {
  const { data, error } = await listSnapshots(projectId);
  if (data == null || error) {
    return [];
  }

  return data.filter((s) => s.metadata == null);
}

export async function removeCorruptedSnapshots(projectId: string) {
  const snapshots = await listCorruptedSnapshots(projectId);
  if (snapshots.length === 0) {
    return;
  }

  return retryOnError(() =>
    supabase.storage
      .from("snapshots")
      .remove(snapshots.map((s) => `${projectId}/${s.name}`))
  );
}
