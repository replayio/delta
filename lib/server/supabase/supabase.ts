import {
  PostgrestResponse,
  PostgrestSingleResponse,
} from "@supabase/supabase-js";
import createClient from "../../initServerSupabase";
import { DeltaErrorCode } from "../../../pages/api/constants";
import { createErrorFromPostgrestError } from "../../../pages/api/utils";

export const supabase = createClient();

export type ResponseError = {
  data: null;
  error: string;
};

export const createError = (error: string): ResponseError => ({
  data: null,
  error,
});

// Runs query and throws if it fails
export async function assertQuery(
  runQuery: () => PromiseLike<PostgrestResponse<unknown>>,
  failureMessage: string
): Promise<void> {
  const result = await retryOnError(runQuery);
  if (result.error) {
    throw createErrorFromPostgrestError(result.error, failureMessage);
  }
}

// Returns a set of records of type <Type> or throws
export async function assertQueryResponse<Type>(
  runQuery: () => PromiseLike<PostgrestResponse<Type>>,
  failureMessage: string
): Promise<Type[]> {
  const result = await retryOnError(runQuery);
  if (result.error || result.data == null) {
    const error = result.error
      ? createErrorFromPostgrestError(result.error, failureMessage)
      : new Error(failureMessage);

    throw error;
  }

  return result.data;
}

// Returns a single record of type <Type> or throws
export async function assertQuerySingleResponse<Type>(
  runQuery: () => PromiseLike<PostgrestSingleResponse<Type>>,
  failureMessage: string
): Promise<Type> {
  const result = await retryOnError(runQuery);
  if (result.error || result.data == null) {
    const error = result.error
      ? createErrorFromPostgrestError(result.error, failureMessage)
      : new Error(failureMessage);

    throw error;
  }

  return result.data;
}

// Returns a single record of type <Type> (or null if none found);
// throws if there is a Postgrest error
export async function assertQuerySingleResponseOrNull<Type>(
  runQuery: () => PromiseLike<PostgrestSingleResponse<Type>>,
  failureMessage: string
): Promise<Type | null> {
  const result = await retryOnError(runQuery);
  if (result.error) {
    const error = result.error
      ? createErrorFromPostgrestError(result.error, failureMessage)
      : new Error(failureMessage);

    throw error;
  }

  return result.data;
}

// Runs query and throws if it fails
export async function assertStorage(
  runQuery: () => PromiseLike<{
    data: unknown | null;
    error: Error | null;
  }>,
  failureMessage: string
): Promise<void> {
  const result = await retryOnError(runQuery);
  if (result.error) {
    throw new Error(`${failureMessage}\n\n${result.error}`);
  }
}

// Returns a single value from of type <Type> or throws
export async function assertStorageValue<Type>(
  runQuery: () => PromiseLike<{
    data: Type | null;
    error: Error | null;
  }>,
  failureMessage: string
): Promise<Type> {
  const result = await retryOnError(runQuery);
  if (result.error || result.data == null) {
    let error = result.error;
    if (!error) {
      error = new Error(failureMessage);
    } else {
      error.message = `${failureMessage}\n\n${error.message}`;
    }

    throw error;
  }

  return result.data;
}

export async function maybeRetry<T>(
  runQuery: () => PromiseLike<T>,
  shouldRetry: (t: T) => boolean
): Promise<T> {
  const callerStackTrace = Error().stack;
  let result: T;
  try {
    result = await runQuery();
    if (!shouldRetry(result)) {
      return result;
    }
    console.error("received error, retrying", result, callerStackTrace);
  } catch (error) {
    console.error("caught error, retrying", error, callerStackTrace);
  }
  try {
    result = await runQuery();
    if (!shouldRetry(result)) {
      return result;
    }
    console.error("received error, giving up", result);
    return result;
  } catch (error) {
    console.error("caught error, giving up", error);
    throw error;
  }
}

export async function retryOnError<T extends { error: any }>(
  runQuery: () => PromiseLike<T>
): Promise<T> {
  return maybeRetry(
    runQuery,
    (result) =>
      !!result.error &&
      result.error.message !==
        "JSON object requested, multiple (or no) rows returned"
  );
}
