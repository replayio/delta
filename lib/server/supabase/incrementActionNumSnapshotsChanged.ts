import { supabase, createError } from "./supabase";
import { getBranchFromProject } from "./branches";
import { getActionFromBranch } from "./actions";

export async function incrementActionNumSnapshotsChanged(
  projectId,
  branchName
) {
  const branch = await getBranchFromProject(projectId, branchName);
  if (branch.error) {
    return createError(`Branch not found for ${branchName}`);
  }
  const action = await getActionFromBranch(branch.data.id);
  if (action.error) {
    return createError(
      `Action not found for ${branchName} and ${
        branch.data.id
      }: ${JSON.stringify(action.error)}`
    );
  }

  return supabase.rpc("snapshots_changed_inc", {
    action_id: action.data.id,
  });
}
