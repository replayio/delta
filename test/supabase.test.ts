import { describe, it, expect } from "vitest";
import orderBy from "lodash/orderBy";
import {
  getActionFromRunId,
  updateAction,
  incrementActionNumSnapshots,
  getAction,
  getActionFromBranch,
} from "../lib/server/supabase/actions";
import omit from "lodash/omit";
import { incrementActionNumSnapshotsChanged } from "../lib/server/supabase/incrementActionNumSnapshotsChanged";
import {
  getSnapshotsFromBranch,
  getSnapshotsForAction,
  getSnapshotFromBranch,
} from "../lib/server/supabase/snapshots";

import { getBranchFromProject } from "../lib/server/supabase/branches";

import { downloadSnapshot } from "../lib/server/supabase/storage";

const projectId = "dcb5df26-b418-4fe2-9bdf-5a838e604ec4";

describe("supabase", () => {
  it("get snapshots", async () => {
    const sn = await getSnapshotsFromBranch(projectId, "main");
    expect(sn.data.length).toBeGreaterThan(0);
  });

  it("get snapshot - warning-stack-collapsed.png", async () => {
    const file = "./playwright/visuals/light/warning-stack-collapsed.png";
    const branchName = "mbudayr/BAC-2510/use-focus-window-to-load-regions";

    const branch = await getBranchFromProject(projectId, branchName);
    const snapshot = await getSnapshotFromBranch(file, projectId, branchName);

    expect(omit(snapshot.data, ["created_at", "id"])).toMatchSnapshot();
    expect(branch.data).toMatchSnapshot();
  });

  it("get action from run_id", async () => {
    const action = await getActionFromRunId("3796486541");
    expect(action.data.id).toEqual("1f11d17c-aa97-4c1f-b312-b1c1954c13f4");
  });

  it("get snapshots for action", async () => {
    const snapshots = await getSnapshotsForAction(
      "7be05e81-8595-4052-a969-3cfaddb44188"
    );
    expect(snapshots.data).toHaveLength(360);
  });

  it("download snapshot", async () => {
    const snapshot = await downloadSnapshot(
      "dcb5df26-b418-4fe2-9bdf-5a838e604ec4/ff8f1f8d9217d9c96e70995292445d33edbd54f411d96b9bdf46733c349da318.png"
    );
    expect(snapshot.data.slice(0, 10)).toEqual("iVBORw0KGg");
  });

  it("can update action status", async () => {
    const actionId = "9ba81dd3-98c4-44a9-a013-5835a1931ae9";
    const action = await updateAction(actionId, { status: "failure" });

    expect(action.data.status).toEqual("failure");

    const action2 = await updateAction(actionId, { status: "success" });

    expect(action2.data.status).toEqual("success");
  });

  it("increments num_snapshots", async () => {
    const actionId = "bb0205c1-0055-40e0-bcad-717ccce77685";
    const action = await getAction(actionId);

    const { num_snapshots } = action.data;
    const { data } = await incrementActionNumSnapshots(actionId);

    const newNumSnapshots = data[0].num_snapshots;
    expect(newNumSnapshots).toEqual(num_snapshots + 1);
  });

  it("increments num_snapshots_changed", async () => {
    const branch = await getBranchFromProject(projectId, "visuals10");
    const action = await getActionFromBranch(branch.data.id);

    const { num_snapshots_changed } = action.data;
    const { data } = await incrementActionNumSnapshotsChanged(
      projectId,
      "visuals10"
    );

    const newNumSnapshots = data[0].num_snapshots_changed;
    expect(newNumSnapshots).toEqual(num_snapshots_changed + 1);
  });
});
