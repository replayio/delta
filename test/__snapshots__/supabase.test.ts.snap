// Vitest Snapshot v1

exports[`supabase > get snapshot - warning-stack-collapsed.png 1`] = `
{
  "action_id": "0cf7f2d3-92c3-4025-87a0-0004b23cc747",
  "changed": null,
  "file": "./playwright/visuals/light/warning-stack-collapsed.png",
  "path": "dcb5df26-b418-4fe2-9bdf-5a838e604ec4/010e1fec31fb30e6771f3ae7b7f7f7053112929b48b9ca612098cb1735c452b8.png",
  "primary_diff_path": "dcb5df26-b418-4fe2-9bdf-5a838e604ec4/8d095ab48e0091f7e0a6ad7959a322f0ad97a392ae34269b2c8a6977a047a3a4.png",
  "primary_num_pixels": null,
  "sha": "010e1fec31fb30e6771f3ae7b7f7f7053112929b48b9ca612098cb1735c452b8",
  "status": "Duplicate",
}
`;

exports[`supabase > get snapshot - warning-stack-collapsed.png 2`] = `
{
  "check_id": "10425425072",
  "comment_id": null,
  "created_at": "2023-01-03T19:21:46+00:00",
  "head_sha": null,
  "id": "c481d064-f58c-4c6b-baa9-af2b888a361c",
  "name": "rebased/mbudayr/BAC-2510/use-focus-window-to-load-regions",
  "pr_number": "8464",
  "pr_title": "Rebased/mbudayr/bac 2510/use focus window to load regions",
  "project_id": "dcb5df26-b418-4fe2-9bdf-5a838e604ec4",
  "status": "closed",
}
`;
